tape/src/main/java/com/squareup/tape/QueueFile.java

```diff
@@ -211,9 +215,34 @@ private Element readElement(int position) throws IOException {
     if (position == 0) return Element.NULL;
     ringRead(position, buffer, 0, Element.HEADER_LENGTH);
     int length = readInt(buffer, 0);
+    checkForElementLengthCorruption(length, position);
     return new Element(position, length);
   }
 
+  /**
+   * If a single element in the queue has a size greater than the length of the content as recorded
+   * in the header, we have a corruption problem. We also check against maxElementSize as no element
+   * should have been inserted with a size greater than that - so if the element's header reports as
+   * such, then again there is corruption.
+   * @param elementLength - Length of the element we are checking.
+   * @param elementPosition - Position of the element's header in the file.
+   * @throws IOException - When there is a file read problem or when there is corruption.
+   */
+  private void checkForElementLengthCorruption(int elementLength, int elementPosition) throws IOException {
+    int fileContentLength = fileLength - QueueFile.HEADER_LENGTH;
+    if (elementLength > fileContentLength || elementLength > maxElementSize) {
+      // Maybe fileContentLength is very small?
+      int debugSize = Math.min(fileContentLength, Math.min(LARGE_ELEMENT_PREVIEW_SIZE, elementLength));
+      byte[] debugBytes = new byte[debugSize];
+      ringRead(elementPosition + Element.HEADER_LENGTH, debugBytes, 0, debugSize);
+      throw new CorruptionException("Possible corruption: Trying to read a Tape element "
+          + "claiming length: " + elementLength + ", which is either a) larger than the current file"
+          + " content length of: " + fileContentLength + ", or larger than the max Tape element size"
+          + " of: " + maxElementSize + ". The first " + debugSize + " bytes of the element are: " +
+          Base64.getEncoder().encodeToString(debugBytes));
```

Comment:
Leaking element contents into an exception message seems like a good way to accidentally expose PII.

Reply:
That's a very good point.

Reply:
All snippets removed.